Source/DafnyLanguageServer/DafnyCodeActions.cs

```diff
@@ -33,22 +33,105 @@ public class DafnyCodeActions {
     if (initialized) {
       return;
     }
     initialized = true;
-    codeActionMap.Add(ErrorID.p_bad_const_initialize_op, (Diagnostic diagnostic, Range range) => ReplacementAction("replace = with :=", diagnostic, range, ":="));
-    codeActionMap.Add(ErrorID.p_abstract_not_allowed, (Diagnostic diagnostic, Range range) => RemoveAction("remove 'abstract'", diagnostic, range));
-    codeActionMap.Add(ErrorID.p_no_leading_underscore, (Diagnostic diagnostic, Range range) => RemoveAction("remove underscore", diagnostic, range));
+    AddRemoveAction(ErrorID.p_duplicate_modifier, "remove duplicate modifier'");
+    AddRemoveAction(ErrorID.p_abstract_not_allowed, "remove 'abstract'");
+    AddRemoveAction(ErrorID.p_no_ghost_for_by_method, "remove 'ghost'");
+    AddRemoveAction(ErrorID.p_ghost_forbidden_default, "remove 'ghost'");
+    AddRemoveAction(ErrorID.p_ghost_forbidden, "remove 'ghost'");
+    AddRemoveAction(ErrorID.p_no_static, "remove 'static'");
+    AddRemoveAction(ErrorID.p_deprecated_attribute, "remove attribute");
+    AddReplaceAction(ErrorID.p_literal_string_required, "replace with empty string", "\"\"");
+    AddRemoveAction(ErrorID.p_no_leading_underscore, "remove underscore");
+    // p_bitvector_too_large -- no code action
+    // p_array_dimension_too_large -- no code action
+    // p_superfluous_exoirt -- TODO - need a code action
+    // p_bad_module_decl -- no code action
+    AddRemoveAction(ErrorID.p_extraneous_comma_in_export, "remove comma");
+    AddReplaceAction(ErrorID.p_top_level_field, "replace 'var' by 'const'", "const"); // also remove entire declaration?
+    // p_bad_datatype_refinement -- what code action
+    AddReplaceAction(ErrorID.p_no_mutable_fields_in_value_types, "replace 'var' by 'const'", "const");
+    AddReplaceAction(ErrorID.p_bad_const_initialize_op, "replace = with :=", ":=");
+    // p_const_is_missing_type_or_init -- no code action
+    AddRemoveAction(ErrorID.p_output_of_function_not_ghost, "remove 'ghost'");
+    AddRemoveAction(ErrorID.p_ghost_function_output_not_ghost, "remove 'ghost'");
+    AddRemoveAction(ErrorID.p_no_new_on_output_formals, "remove 'new'");
+    AddRemoveAction(ErrorID.p_no_nameonly_on_output_formals, "remove 'nameonly'");
+    AddRemoveAction(ErrorID.p_no_older_on_output_formals, "remove 'older'");
+    // p_var_decl_must_have_type -- no code action
+    // p_no_init_for_var_field -- remove entire initiallizer
+    // p_datatype_formal_is_not_id -- no code action - perhaps remove the name and the colon
+    AddRemoveAction(ErrorID.p_nameonly_must_have_parameter_name, "remove 'nameonly'");
+    AddReplaceAction(ErrorID.p_should_be_yields_instead_of_returns, "replace 'returns' by 'yields'", "yields");
+    AddRemoveAction(ErrorID.p_type_parameter_variance_forbidden, "remove type parameter variance");
+
+    AddReplacementActions(ErrorID.p_unexpected_type_characteristic, new Tuple<String, String>[]{
+      Tuple.Create("replace with '=='", "=="),
```

Comment:
Don't you want to insert instead of replacing?

Reply:
In this case there is some token there that is illegal as a type characteristic -- so the action is to replace it.

Reply:
Oh yeah, I see it's triggered only when given digits which are not 0 or 00, makes sense